### PR TITLE
gettext: Update to version 0.22.4

### PR DESCRIPTION
**Pull request by tifrueh**
* Update version number
* Update checksums
* Update to new build process
  * Build the whole of gettext in gettext and gettext-tools-libs, but only install the correct files for either port
  * This is due to 'gettext-tools' not being buildable without building 'gettext-runtime' first
  * Add ncurses as build dependency for gettext and gettext-tools-libs because libtextstyle is built again when building gettext and gettext-tools-libs
* Remove deactivate hack

Closes: https://trac.macports.org/ticket/68030

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 14.1.1 23B81 arm64
Xcode 15.0.1 15A507

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by ryandesign** [comment]
It would have been better to update #21372 so that the conversation is not split into multiple places.

**Comment by tifrueh** [comment]
Oh, I'm sorry. I thought I'd have to open a new one for a different version.
What are the preferred next steps from here on out?

File: devel/gettext/Portfile [hunk 1]
```diff
@@ -182,29 +177,19 @@ subport ${name}-tools-libs {
...
+     destroot.dir           ${worksrcpath}/${name}-tools
+
     post-destroot {
         # Remove the executables and the libraries provided
         # by the runtime. Just leave the libraries and headers
         # for the tools.
-        delete {*}[glob ${destroot}${prefix}/*/libintl.*] \
-            {*}[glob ${destroot}${prefix}/bin/*] \
+        delete {*}[glob ${destroot}${prefix}/bin/*] \
             {*}[glob ${destroot}${prefix}/share/*]
```
Comment (ryandesign) [inline devel/gettext/Portfile:189]: You have removed the portfile code that deleted libraries but you haven't removed the comment that says libraries get deleted.

File: devel/gettext/Portfile [hunk 2]
```diff
@@ -182,29 +177,19 @@ subport ${name}-tools-libs {
     # https://savannah.gnu.org/bugs/?58669
     configure.args-append   --disable-libasprintf
 
+    # Install from the tools subpackage, so that only its files
+    # are installed
```
Comment (ryandesign) [inline devel/gettext/Portfile:181]:
```suggestion
    # are installed.
```

File: devel/gettext/Portfile [hunk 3]
```diff
@@ -256,6 +236,10 @@ if {${subport} eq ${name}} {
...
```
Comment (ryandesign) [inline devel/gettext/Portfile:240]:
```suggestion
    # are installed.
```

**Review by ryandesign — COMMENTED** [review]
It is of course critical that we update to gettext 0.22.x since it fixes crashes on Sonoma. I have not tested this PR. I am annoyed that, if I understand correctly, the changes that had to be made to the portfile increase the build time. I have voiced my annoyance with the developers of gettext and they are uninterested and I have no remaining energy to argue with them further.

**Comment by tifrueh** [comment]
Damn. I just read your update on https://savannah.gnu.org/bugs/?58669 and the developer's response to it. It really seems to me that building some parts multiple times is the only way to do it in our case; The developer suggests building the whole package once and then distributing all the different files to their corresponding package, which (to my knowledge) can't be accomplished with macports. For cases where this method is not possible, he says:

> For distros that are not of this type, I have documented the instructions and restrictions in the PACKAGING files. In this case, gettext-runtime needs to be built twice. But it is a relatively small package.

and

> In your first comment, you wrote:
> 
>\> I want gettext-tools to use the libintl that I already built and installed previously via gettext-runtime.
> 
> But this will not be done. It would be too complicated and too fragile.

So using the same libintl twice really does seem out of the question.

And because we split gettext-tools into gettext-tools-libs and gettext (I'm still a tad confused about that, what is the reason for it again?) we need to build gettext-runtime an additional time.

**Comment by tifrueh** [comment]
Oh, and I just noticed that I probably made a mistake when updating which files should be deleted in the post-destroot section of one or more subports, some of them might be missing as a result. I'll fix that as soon as possible and I'll incorporate your reviews on the comments as well.

**Comment by pmetzger** [comment]
@ryandesign Any opinion on the latest version?

**Comment by pmetzger** [comment]
Howdy, all. If @ryandesign has an opinion that he expresses soon, we'll of course follow his requests. If he does not, then I think we should simply commit this as is.

**Comment by ryandesign** [comment]
> And because we split gettext-tools into gettext-tools-libs and gettext (I'm still a tad confused about that, what is the reason for it again?)

Explained in this commit message: baf0c8bd508912bfedaf58c40dba1a402ffdf2ed

**Comment by ryandesign** [comment]
> Any opinion on the latest version?

First, thank you so much to @tifrueh for doing this and sorry it's taking so long.

What has delayed me is that since there have been changes in the build system, I wanted to verify that each subport still installs the files it's supposed to. What I need to do is build each subport on my system and diff the file list against the old 0.21.1 subports. And I've been putting off doing that since I know it'll take awhile.

And as with any gettext update, it's critical that it builds on all systems and is correct. Everything depends on gettext so if we commit something broken, that will cause buildbot builds of most subsequent ports to fail, and rescheduling those builds once the gettext problem is fixed is extra work that I want to avoid. Since GitHub CI only checks recent macOS versions, I should at least try to build it on a couple older systems like 10.5 and 10.7 or whichever one uses the oldest Xcode clang. And whenever this PR is merged, it should be done when all or at least most of the buildbot workers are idle, and the buildbot progress should be monitored to verify that all builds succeed so that if they fail, the problem can be investigated and fixed ASAP. (I would probably do a graceful shutdown on any worker where a gettext build failed, to prevent other builds from running until the problem is fixed.) At the moment, there are long backlogs on at least some of the workers.

I'll want to include a fix for https://trac.macports.org/ticket/69195 too, which I need to verify first also.

**Comment by ryandesign** [comment]
I built gettext and all the subports successfully on macOS 12, 10.11, 10.10, 10.9, 10.8, 10.7, and 10.5 without the universal variant and on macOS 12 with the universal variant, all on x86_64.

I verified that the problem with clang 6xx remains on 10.9 when the blacklist is removed and that my patch fixes it.

I diff'd the file list of the 0.21.1 and 0.22.4 versions of all the (sub)ports on 10.9 and I'm happy with them; I don't see any files missing from any of them.

So thank you again! I'll merge this in a minute.

**Comment by tifrueh** [comment]
My pleasure, I'm glad to have been of help.

**Pull request merged**